qiskit/transpiler/passes/synthesis/qubit_tracker.py

```diff
@@ -25,108 +25,95 @@ class QubitTracker:
+        for q in qubits:
+            self.enabled[q] = False
+
+    def enable(self, qubits):
+        """Enables using the given qubits."""
+        for q in qubits:
+            self.enabled[q] = True
+
+    def num_clean(self, ignored_qubits):
+        """Returns the number of enabled clean qubits, ignoring the given qubits."""
+        count = 0
+        for q in ignored_qubits:
+            self.ignored[q] = True
+        for q in range(self.num_qubits):
+            if (not self.ignored[q]) and self.enabled[q] and self.state[q]:
+                count += 1
+        for q in ignored_qubits:
+            self.ignored[q] = False
+        return count
+
+    def num_dirty(self, ignored_qubits):
+        """Returns the number of enabled dirty qubits, ignoring the given qubits."""
+        count = 0
+        for q in ignored_qubits:
+            self.ignored[q] = True
+        for q in range(self.num_qubits):
+            if (not self.ignored[q]) and self.enabled[q] and not self.state[q]:
+                count += 1
+        for q in ignored_qubits:
+            self.ignored[q] = False
+        return count
+
+    def borrow(self, num_qubits: int, ignored_qubits: Iterable[int] | None = None) -> list[int]:
+        """Get ``num_qubits`` enabled qubits, excluding ``ignored_qubits`` and prioritizing
+        clean qubits."""
+        res = []
+        for q in ignored_qubits:
+            self.ignored[q] = True
+        for q in range(self.num_qubits):
+            if (not self.ignored[q]) and self.enabled[q] and self.state[q]:
+                res.append(q)
+        for q in range(self.num_qubits):
+            if (not self.ignored[q]) and self.enabled[q] and not self.state[q]:
+                res.append(q)
+        for q in ignored_qubits:
+            self.ignored[q] = False
+        return res[:num_qubits]
+
+    def copy(self) -> "QubitTracker":
+        """Copies the qubit tracker."""
+        tracker = QubitTracker(self.num_qubits)
+        tracker.state = self.state.copy()
+        tracker.enabled = self.enabled.copy()
+        # no need to copy the scratch space (ignored)
+        return tracker
+
+    def replace_state(self, other: "QubitTracker", qubits):
+        """Replaces the state of the given qubits by their state in the ``other`` tracker."""
+        for q in qubits:
+            self.state[q] = other.state[q]
 
     def __str__(self) -> str:
-        return (
-            f"QubitTracker({len(self.qubits)}, clean: {self.num_clean()}, dirty: {self.num_dirty()})"
-            + f"\n\tclean: {self.clean}"
-            + f"\n\tdirty: {self.dirty}"
-        )
+        """Pretty-prints qubit states."""
+        out = ""
+        for q in range(self.num_qubits):
+            out += str(q) + ": "
+            if not self.enabled[q]:
+                out += "_"
+            elif self.state[q]:
+                out += "0"
+            else:
+                out += "*"
+            out += "; "
+        return out
```

Comment:
Done in [60e7456](https://github.com/Qiskit/qiskit/pull/13240/commits/60e745607d27f01a16d9c8f80061bdb91ad7ccb1). Though this might become obsolete after porting to Rust.